# DATA CLEANING:

In [1]:
import pandas as pd 
import numpy as np
df = pd.read_csv(r'D:\Huy\Daily Work\Financial Analysis Daily Practice\Data Projecct\Sewing_FS24.csv')


In [2]:
print(df.head())

   Company_ID  FS year Cash & Bank Inventory Accounts Receivable Investment  \
0           1     2024      10,150       590              53,429       -      
1           2     2024       6,150     5,912               9,552       -      
2           3     2024     101,876   147,476             115,182       -      
3           4     2024       4,714     4,049             321,473       -      
4           5     2024      20,598    11,230              51,878       -      

  Total Current Assets Total Assets Accounts Payable  \
0               64,679       66,248            1,598   
1               21,767       63,571            1,197   
2              365,284      495,920          122,764   
3              337,746      342,685          336,846   
4               85,995       86,715            9,683   

  Short- term loans and finance lease liabilities  ... Total Equity  \
0                                            -     ...      -42,103   
1                                           8,

In [3]:
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1594 entries, 0 to 1593
Data columns (total 25 columns):
 #   Column                                           Non-Null Count  Dtype 
---  ------                                           --------------  ----- 
 0   Company_ID                                       1594 non-null   int64 
 1   FS year                                          1594 non-null   int64 
 2   Cash & Bank                                      1594 non-null   object
 3   Inventory                                        1594 non-null   object
 4   Accounts Receivable                              1594 non-null   object
 5   Investment                                       1594 non-null   object
 6   Total Current Assets                             1594 non-null   object
 7   Total Assets                                     1594 non-null   object
 8   Accounts Payable                                 1594 non-null   object
 9   Short- term loans and finance lease liabi

,Company_ID,FS year,Cash & Bank,Inventory,Accounts Receivable,Investment,Total Current Assets,Total Assets,Accounts Payable,Short- term loans and finance lease liabilities,...,Total Equity,Total Liabilities & Equity,Net Revenue,COSG,Gross Profit,Profit from operations,Net profit before tax,Net profit after tax,Tangible Net Worth,Interest expenses
0,1,2024,"10,150",590,"53,429",-,"64,679","66,248","1,598",-,...,"-42,103","66,248","61,057",-63319.00,"-2,262","-14,353","-14,353","-14,353","-43,223",NaN
1,2,2024,"6,150","5,912","9,552",-,"21,767","63,571","1,197","8,029",...,"51,338","63,571","70,771",-59378.00,"11,393","-5,863","-4,821","-4,821","51,201",440
2,3,2024,"101,876","147,476","115,182",-,"365,284","495,920","122,764",-,...,"221,143","495,920","727,057",-654719.00,"72,338","42,684","45,668","39,269","220,928",NaN
3,4,2024,"4,714","4,049","321,473",-,"337,746","342,685","336,846",-,...,"-80,236","342,685","112,761",-108526.00,"4,236",264,263,151,"-80,236",-
4,5,2024,"20,598","11,230","51,878",-,"85,995","86,715","9,683",12,...,"65,625","86,715","165,422",-93609.00,"71,814","33,270","32,502","26,001","65,625",-


In [4]:
#Drop unnecessary columns that do not contribute to the analysis.
keep_cols = [
    'Total Current Assets', 'Total current Liabilities', 'Cash & Bank',
    'Inventory', 'Accounts Receivable',
    'Total Assets', 'Total Liabilities', 'Total Equity',
    'Net Revenue', 'COSG', 'Interest expenses', 'Net profit after tax'
]

df_clean = df[keep_cols]

In [5]:
df_clean = df_clean.rename(columns={
    'Total Current Assets': 'T_Current_Assets','Total current Liabilities':'T_Current_Liabilities','Cash & Bank':'Cash','Accounts Receivable':'Accounts_Receivable',
    'Total Assets':'T_Assets','Total Liabilities':'T_Liabilities','Total Equity':'T_Equity','Net Revenue':'Net_Revenue','Interest expenses':'IE',
    'Net profit after tax':'NPAT','COSG':'COGS'
})

In [6]:
df_clean = df_clean.replace({' ': '', ',': ''}, regex=True)

df_clean = df_clean.replace('-', '0', regex=True)

In [7]:
print(df_clean.dtypes)

T_Current_Assets         object
T_Current_Liabilities    object
Cash                     object
Inventory                object
Accounts_Receivable      object
T_Assets                 object
T_Liabilities            object
T_Equity                 object
Net_Revenue              object
COGS                     object
IE                       object
NPAT                     object
dtype: object


In [8]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1594 entries, 0 to 1593
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   T_Current_Assets       1594 non-null   object
 1   T_Current_Liabilities  1594 non-null   object
 2   Cash                   1594 non-null   object
 3   Inventory              1594 non-null   object
 4   Accounts_Receivable    1594 non-null   object
 5   T_Assets               1594 non-null   object
 6   T_Liabilities          1594 non-null   object
 7   T_Equity               1594 non-null   object
 8   Net_Revenue            1594 non-null   object
 9   COGS                   1594 non-null   object
 10  IE                     569 non-null    object
 11  NPAT                   1594 non-null   object
dtypes: object(12)
memory usage: 149.6+ KB


In [9]:
col_name = [
    'T_Current_Assets','T_Current_Liabilities','Cash','Inventory',
    'Accounts_Receivable','T_Assets','T_Liabilities','T_Equity',
    'Net_Revenue','COGS','IE','NPAT'
]

for col in col_name:
    df_clean[col] = (
        df_clean[col]
        .astype(str)
        .str.replace('.', '', regex=False) # remove dots safely
        .str.replace('-','', regex=False)   
        .str.strip()
    )

    df_clean[col] = pd.to_numeric(df_clean[col], errors='coerce')

    

In [10]:
# Prevent zero / near-zero denominators
df_clean['Inventory'] = df_clean['Inventory'].mask(df_clean['Inventory'] < 1, np.nan)
df_clean['COGS'] = df_clean['COGS'].mask(df_clean['COGS'] < 1, np.nan)
df_clean['T_Current_Liabilities'] = df_clean['T_Current_Liabilities'].mask(df_clean['T_Current_Liabilities'] < 1, np.nan)
df_clean['Accounts_Receivable'] = df_clean['Accounts_Receivable'].mask(df_clean['Accounts_Receivable'] < 1, np.nan)
df_clean['T_Equity'] = df_clean['T_Equity'].mask(df_clean['T_Equity'] < 1, np.nan)

In [11]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1594 entries, 0 to 1593
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   T_Current_Assets       1594 non-null   int64  
 1   T_Current_Liabilities  1585 non-null   float64
 2   Cash                   1594 non-null   int64  
 3   Inventory              1447 non-null   float64
 4   Accounts_Receivable    1572 non-null   float64
 5   T_Assets               1594 non-null   int64  
 6   T_Liabilities          1594 non-null   int64  
 7   T_Equity               1594 non-null   int64  
 8   Net_Revenue            1594 non-null   int64  
 9   COGS                   1551 non-null   float64
 10  IE                     569 non-null    float64
 11  NPAT                   1594 non-null   int64  
dtypes: float64(5), int64(7)
memory usage: 149.6 KB


Interest expense in the raw dataset has too many missing values, even though it is a very important figure for calculating interest coverage. I had to drop this column.

In [12]:
df_clean = df_clean.drop(columns=['IE'])
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1594 entries, 0 to 1593
Data columns (total 11 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   T_Current_Assets       1594 non-null   int64  
 1   T_Current_Liabilities  1585 non-null   float64
 2   Cash                   1594 non-null   int64  
 3   Inventory              1447 non-null   float64
 4   Accounts_Receivable    1572 non-null   float64
 5   T_Assets               1594 non-null   int64  
 6   T_Liabilities          1594 non-null   int64  
 7   T_Equity               1594 non-null   int64  
 8   Net_Revenue            1594 non-null   int64  
 9   COGS                   1551 non-null   float64
 10  NPAT                   1594 non-null   int64  
dtypes: float64(4), int64(7)
memory usage: 137.1 KB


Because the dataset is messy and contains invalid values (less than or equal to 0), I applied filtering conditions to remove those rows.

In [13]:
condition_garbage = (df_clean['T_Assets'] <= 0) | (df_clean['Net_Revenue'] <= 0)

df_garbage = df_clean[condition_garbage]

df_clean = df_clean[~condition_garbage].copy()

Another issue was records where total assets did not equal the sum of total liabilities and total equity, so I applied filtering conditions to remove those rows.

In [14]:
# Set tolerance deviation to 1000 VND
tolerance = 1000 

# Check the deviation of value
df_clean['Balance_Diff'] = abs(df_clean['T_Assets'] - (df_clean['T_Liabilities'] + df_clean['T_Equity']))

# Remove invalid rows from the dataset
df_clean = df_clean[df_clean['Balance_Diff'] <= tolerance].copy()

In [15]:
#Start with Efficiency Ratios#
df_clean['Inv_Turn'] = df_clean['COGS']/df_clean['Inventory'].replace(0, np.nan)
df_clean['DSI'] = 365 * df_clean['Inventory'] / df_clean['COGS'].replace(0, np.nan)
#Profitability Ratios"
df_clean['ROA'] = df_clean['NPAT']/df_clean['T_Assets'].replace(0, np.nan)
df_clean['ROE'] = df_clean['NPAT']/df_clean['T_Equity'].replace(0, np.nan)
df_clean['Cur_Ratio'] = df_clean['T_Current_Assets']/df_clean['T_Current_Liabilities'].replace(0, np.nan)
df_clean['Q_Ratio'] = (df_clean['T_Current_Assets']-df_clean['Inventory'])/df_clean['T_Current_Liabilities'].replace(0, np.nan)
df_clean['Cas_Ratio'] = df_clean['Cash']/ df_clean['T_Current_Liabilities'].replace(0, np.nan)
df_clean['Receivable_Turn'] = df_clean['Net_Revenue']/df_clean['Accounts_Receivable'].replace(0, np.nan)
#Leverage Ratios"
df_clean['DE'] = df_clean['T_Liabilities']/df_clean['T_Equity'].replace(0, np.nan)
df_clean['Debt_Ratio'] = df_clean['T_Liabilities']/df_clean['T_Assets'].replace(0, np.nan)

In [16]:
df_clean = df_clean.replace([np.inf, -np.inf], np.nan)

In [17]:
missing_count = df_clean.isna().sum()
print(missing_count)

T_Current_Assets          0
T_Current_Liabilities     6
Cash                      0
Inventory                96
Accounts_Receivable       9
T_Assets                  0
T_Liabilities             0
T_Equity                  0
Net_Revenue               0
COGS                      8
NPAT                      0
Balance_Diff              0
Inv_Turn                 97
DSI                      97
ROA                       0
ROE                       0
Cur_Ratio                 6
Q_Ratio                  97
Cas_Ratio                 6
Receivable_Turn           9
DE                        0
Debt_Ratio                0
dtype: int64


Since the Inventory Turnover and Receivables Turnover, DSI, Current Ratios, Quick Ratios, Cash Ratios columns contain only a small number of missing values relative to the entire dataset, I will drop those rows.

In [18]:
df_clean = df_clean.dropna()

The DSI formula is correctly applied; however, extremely high values arise when the cost of goods sold (COGS) is very small, leading to inflated ratios. To address this, observations with near-zero denominators are treated and extreme values are capped to ensure meaningful interpretation.

In [19]:
df_clean['DSI'] = df_clean['DSI'].clip(upper=365)

In [20]:
df_clean.describe()

,T_Current_Assets,T_Current_Liabilities,Cash,Inventory,Accounts_Receivable,T_Assets,T_Liabilities,T_Equity,Net_Revenue,COGS,...,Inv_Turn,DSI,ROA,ROE,Cur_Ratio,Q_Ratio,Cas_Ratio,Receivable_Turn,DE,Debt_Ratio
count,1.238000e+03,1.238000e+03,1.238000e+03,1.238000e+03,1.238000e+03,1.238000e+03,1.238000e+03,1.238000e+03,1.238000e+03,1.238000e+03,...,1.238000e+03,1238.000000,1238.000000,1238.000000,1238.000000,1238.000000,1238.000000,1238.000000,1238.000000,1238.000000
mean,1.753699e+05,1.289546e+05,2.637855e+04,5.488710e+04,8.030218e+04,2.846800e+05,1.558075e+05,1.288748e+05,3.247338e+05,2.754762e+07,...,1.519184e+04,1.818243,0.055713,0.368873,78.450198,18.712183,2.201840,246.122728,7.166146,0.575760
std,5.023343e+05,3.706779e+05,9.613229e+04,1.708401e+05,2.748402e+05,8.398515e+05,4.720938e+05,4.489635e+05,8.978823e+05,7.412621e+07,...,2.636076e+05,13.203522,0.098972,2.921337,2609.182909,540.920677,41.987095,7227.547508,48.039184,0.267300
min,9.520000e+02,2.000000e+00,3.000000e+00,1.000000e+00,1.000000e+00,1.711000e+03,2.600000e+01,6.000000e+01,1.110000e+02,1.110000e+04,...,2.037135e-01,0.000041,0.000000,0.000000,0.065910,0.023579,0.000069,0.002692,0.000719,0.000719
25%,2.425750e+04,1.596975e+04,1.875750e+03,3.265500e+03,6.899000e+03,3.370900e+04,1.722325e+04,8.785000e+03,4.322375e+04,3.719125e+06,...,3.329777e+02,0.177063,0.004823,0.015118,1.007932,0.594186,0.050348,2.864669,0.552020,0.355678
50%,5.534000e+04,3.946000e+04,7.037500e+03,1.231600e+04,1.958450e+04,8.445100e+04,4.381700e+04,2.717750e+04,1.025360e+05,9.001750e+06,...,8.608582e+02,0.423996,0.022484,0.066050,1.310419,0.970587,0.179079,5.620795,1.624984,0.619045
75%,1.379095e+05,9.669725e+04,1.932250e+04,3.767225e+04,5.870525e+04,2.291300e+05,1.096025e+05,9.311750e+04,2.627048e+05,2.267638e+07,...,2.061416e+03,1.096171,0.074370,0.200377,2.172034,1.653211,0.554222,10.939322,4.149340,0.805789
max,9.551569e+06,7.066894e+06,2.489488e+06,3.217896e+06,5.242607e+06,1.926585e+07,9.989509e+06,9.276342e+06,1.732583e+07,1.545085e+09,...,8.866200e+06,365.000000,1.593533,90.361564,91796.000000,18979.500000,1376.807692,253903.000000,1386.928339,1.026128


As we can see, the distribution reflects a typical economic reality: the dense cluster on the left represents the majority of SMEs in Vietnam Sewing Industry, while the heavy right tail is driven by a small number of large corporations with assets or revenue significantly higher than the mean.

To mitigate the distortion caused by outliers, the Interquartile Range (IQR) method was applied with a sensitivity coefficient of $k = 3.0$. This approach effectively filters out extreme anomalies while preserving the data of high-performing, 'excellent' companies that represent the upper tier of the industry.

In [21]:
def clean_financial_outliers_fixed(df_clean):
    df_out = df_clean.copy()
    # Create the first mask where every row is True
    is_not_outlier = pd.Series(True, index=df_out.index)
    
    for col in [('Cur_Ratio'), ('Q_Ratio'), ('Cas_Ratio'), 
                   ('Inv_Turn'), ('Receivable_Turn'), ('DSI'),
                   ('ROA'), ('ROE')]:
        
        if col in df_out.columns:
            Q1 = df_out[col].quantile(0.25)
            Q3 = df_out[col].quantile(0.75)
            IQR = Q3 - Q1
            
            # Updating mask: So if any row does not achive the additional the condition will be removed
            condition = (df_out[col] >= Q1 - 3.0*IQR) & (df_out[col] <= Q3 + 3.0*IQR)
            is_not_outlier &= condition
            
    return df_out[is_not_outlier]


df_final_clean = clean_financial_outliers_fixed(df_clean)

In [22]:
df_final_clean.describe()

,T_Current_Assets,T_Current_Liabilities,Cash,Inventory,Accounts_Receivable,T_Assets,T_Liabilities,T_Equity,Net_Revenue,COGS,...,Inv_Turn,DSI,ROA,ROE,Cur_Ratio,Q_Ratio,Cas_Ratio,Receivable_Turn,DE,Debt_Ratio
count,8.560000e+02,8.560000e+02,8.560000e+02,8.560000e+02,8.560000e+02,8.560000e+02,8.560000e+02,8.560000e+02,8.560000e+02,8.560000e+02,...,856.000000,856.000000,856.000000,856.000000,856.000000,856.000000,856.000000,856.000000,856.000000,856.000000
mean,1.888184e+05,1.510804e+05,2.465518e+04,5.757749e+04,9.060630e+04,3.105347e+05,1.771578e+05,1.333776e+05,3.695199e+05,3.212764e+07,...,1254.056883,0.776825,0.041020,0.117743,1.595976,1.144525,0.303822,7.179841,4.010592,0.597532
std,5.042322e+05,4.208707e+05,6.655431e+04,1.583128e+05,2.852545e+05,9.012075e+05,5.323447e+05,4.233586e+05,9.197622e+05,8.066833e+07,...,1359.633185,0.787041,0.052056,0.146964,0.952198,0.808833,0.381661,6.518309,8.753983,0.233424
min,3.806000e+03,1.755000e+03,1.900000e+01,1.260000e+02,4.960000e+02,3.806000e+03,1.755000e+03,1.290000e+02,2.145000e+03,1.404000e+05,...,94.898987,0.051938,0.000000,0.000000,0.102360,0.068032,0.000560,0.046594,0.021861,0.021393
25%,2.947950e+04,2.098550e+04,1.995750e+03,5.101500e+03,1.021325e+04,4.216350e+04,2.250850e+04,9.973500e+03,4.965700e+04,4.397175e+06,...,347.899502,0.229493,0.004147,0.014663,1.023393,0.622876,0.048497,2.891858,0.716496,0.417417
50%,6.463500e+04,4.874400e+04,7.451500e+03,1.494750e+04,2.570500e+04,9.337150e+04,5.312600e+04,3.043700e+04,1.242210e+05,1.055860e+07,...,758.159683,0.481441,0.017551,0.057647,1.278039,0.918027,0.149975,5.197646,1.748797,0.636204
75%,1.551630e+05,1.117435e+05,2.049575e+04,4.419200e+04,6.848400e+04,2.610262e+05,1.238470e+05,1.091942e+05,3.466308e+05,3.043832e+07,...,1590.465113,1.049160,0.062756,0.169801,1.870239,1.367699,0.412894,9.253283,3.719557,0.788108
max,9.551569e+06,7.066894e+06,1.004605e+06,3.217896e+06,5.242607e+06,1.926585e+07,9.989509e+06,9.276342e+06,1.732583e+07,1.545085e+09,...,7027.612110,3.846195,0.275811,0.751056,5.448177,4.725508,1.979733,34.340767,113.112745,1.010522


In [23]:
df_final_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 856 entries, 1 to 1593
Data columns (total 22 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   T_Current_Assets       856 non-null    int64  
 1   T_Current_Liabilities  856 non-null    float64
 2   Cash                   856 non-null    int64  
 3   Inventory              856 non-null    float64
 4   Accounts_Receivable    856 non-null    float64
 5   T_Assets               856 non-null    int64  
 6   T_Liabilities          856 non-null    int64  
 7   T_Equity               856 non-null    int64  
 8   Net_Revenue            856 non-null    int64  
 9   COGS                   856 non-null    float64
 10  NPAT                   856 non-null    int64  
 11  Balance_Diff           856 non-null    int64  
 12  Inv_Turn               856 non-null    float64
 13  DSI                    856 non-null    float64
 14  ROA                    856 non-null    float64
 15  ROE       

## In financial data analysis, "dirty data" is significantly more damaging than "limited data." 
* The Risk of the Full Dataset (1,500 rows): While providing a larger sample size, extreme anomalies (e.g., Cur_Ratio = 91,796) and contain accounting discrepancies where Total Assets does not equal Total Sum of Total Equity and Liabilities, would heavily distort sector averages. This would lead to skewed visualizations (Heatmaps) and biased statistical conclusions.

* The Benefit of the Refined Dataset (856 rows): These records represent "economic reality." By filtering for logical financial ranges, we ensure that correlation coefficients and trends are reliable and statistically significant.

I will categorize company size based on total assets, as the sewing industry relies heavily on physical resources such as machinery and inventory to operate. While total assets may not directly measure production capacity, they can serve as a reasonable proxy for a company’s operational scale in this industry.

In [24]:
df_rank = df_final_clean.copy()

ranks = []
for index, row in df_rank.iterrows(): 
    assets = row['T_Assets']
    if assets > 500000: ranks.append('LARGE')
    elif 200000 < assets < 500000: ranks.append('BIG')
    else: ranks.append('SME')   

df_final_clean['Rank'] = ranks

In [25]:
df_final_clean['Company_ID'] = np.arange(1, len(df_final_clean) + 1)

In [26]:
df_final_clean['DE'].describe()

count    856.000000
mean       4.010592
std        8.753983
min        0.021861
25%        0.716496
50%        1.748797
75%        3.719557
max      113.112745
Name: DE, dtype: float64